src/Controls/src/Core/Element.cs

```diff
@@ -97,7 +98,110 @@ public string StyleId
 			}
 		}
 
-		internal virtual IReadOnlyList<Element> LogicalChildrenInternal => EmptyChildren;
+		// Leaving this internal for now.
+		// If users want to add/remove from this they can use
+		// AddLogicalChildren and RemoveLogicalChildren on the respective control
+		// if available.
+		//
+		// Ultimately I don't think we'll need these to be virtual but some controls (layout)
+		// are going to take a more focused effort so I'd rather just do that in a 
+		// separate PR. I don't think there's ever a scenario where a subclass needs
+		// to replace the backing store.
+		// If everyone just uses AddLogicalChildren and RemoveLogicalChildren
+		// and then overrides OnChildAdded/OnChildRemoved
+		// that should be sufficient
+		internal IReadOnlyList<Element> LogicalChildrenInternal
+		{
+			get
+			{
+				SetupChildren();
+				return _logicalChildrenReadonly;
+			}
+		}
+
+		private protected virtual IList<Element> LogicalChildrenInternalBackingStore
+		{
+			get
+			{
+				_internalChildren ??= new List<Element>();
+				return _internalChildren;
+			}
+		}
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		[Obsolete("Do not use! This is to be removed! Just used by Hot Reload! To be replaced with IVisualTreeElement!")]
+		public ReadOnlyCollection<Element> LogicalChildren =>
+			new ReadOnlyCollection<Element>(new TemporaryWrapper(LogicalChildrenInternal));
+
+		void SetupChildren()
+		{
+			_logicalChildrenReadonly ??= new ReadOnlyCollection<Element>(LogicalChildrenInternalBackingStore);
+		}
+
+		internal void InsertLogicalChildInternal(int index, Element element)
+		{
+			if (element is null)
+			{
+				return;
+			}
+
+			SetupChildren();
+
+			LogicalChildrenInternalBackingStore.Insert(index, element);
+			OnChildAdded(element);
+		}
+
+		internal void AddLogicalChildInternal(Element element)
+		{
+			if (element is null)
+			{
+				return;
+			}
+
+			SetupChildren();
+
+			LogicalChildrenInternalBackingStore.Add(element);
+			OnChildAdded(element);
+		}
+
+		internal bool RemoveLogicalChildInternal(Element element)
+		{
+			if (element is null)
+			{
+				return false;
+			}
+
+			if (LogicalChildrenInternalBackingStore is null || !LogicalChildrenInternalBackingStore.Contains(element))
+				return false;
+
+			var oldLogicalIndex = LogicalChildrenInternalBackingStore.IndexOf(element);
+			RemoveLogicalChildInternal(element, oldLogicalIndex);
+
+			return true;
+		}
+
+		internal void ClearLogicalChildren()
+		{
+			if (LogicalChildrenInternalBackingStore is null)
+				return;
+
+			if (LogicalChildrenInternal == EmptyChildren)
+				return;
+
+			// Reverse for-loop, so children can be removed while iterating
+			for (int i = LogicalChildrenInternalBackingStore.Count - 1; i >= 0; i--)
+			{
+				RemoveLogicalChildInternal(LogicalChildrenInternalBackingStore[i], i);
+			}
+		}
+
+		bool RemoveLogicalChildInternal(Element element, int oldLogicalIndex)
```

Comment:
If this is internal then others can use it. I see ContextFlyout can benefit as it both does a lookup in RemoveAt as well as an Index Of in remove. So that can skip the search.